File: mili-server/src/main/java/fun/bm/mili/command/counter/CounterCommand.java
package fun.bm.mili.command.counter;

import fun.bm.mili.command.counter.sub.DisplayCommand;
import fun.bm.mili.command.counter.sub.ResetCommand;
import fun.bm.mili.command.counter.sub.ToggleCommand;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.command.CommandSender;
import org.leavesmc.leaves.command.CommandContext;
import org.leavesmc.leaves.command.RootNode;
import org.leavesmc.leaves.util.HopperCounter;

public class CounterCommand extends RootNode {

    private static final String PERM_BASE = "mili.commands.counter";

    public CounterCommand() {
        super("counter", PERM_BASE);
        children(ToggleCommand::new, ResetCommand::new, DisplayCommand::new);
    }

    @Override
    public boolean execute(CommandContext context) {
        context.getSender().sendMessage(
            Component.join(
                JoinConfiguration.noSeparators(),
                Component.text("Hopper Counter: ", NamedTextColor.GRAY),
                Component.text(
                    HopperCounter.isEnabled(),
                    HopperCounter.isEnabled() ? NamedTextColor.AQUA : NamedTextColor.GRAY
                )
            )
        );
        return true;
    }

    public static boolean hasPermission(CommandSender sender, String... subcommand) {
        String suffix = String.join(".", subcommand);
        if (suffix.isEmpty()) {
            return sender.hasPermission(PERM_BASE);
        }
        return sender.hasPermission(PERM_BASE) || sender.hasPermission(PERM_BASE + "." + suffix);
    }
}

File: mili-server/src/main/kotlin/fun/bm/mili/command/counter/CounterCommand.kt
@file:JvmName("CounterCommandKt")

package fun.bm.mili.command.counter

import com.mojang.brigadier.exceptions.CommandSyntaxException
import fun.bm.mili.command.counter.sub.DisplayCommand
import fun.bm.mili.command.counter.sub.ResetCommand
import fun.bm.mili.command.counter.sub.ToggleCommand
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.JoinConfiguration
import net.kyori.adventure.text.format.NamedTextColor
import org.leavesmc.leaves.command.CommandContext
import org.leavesmc.leaves.command.RootNode
import org.leavesmc.leaves.util.HopperCounter

class CounterCommand : RootNode("counter", "mili.commands.counter") {

    private val permBase = "mili.commands.counter"

    init {
        children(
            ToggleCommand(this),
            ResetCommand(this),
            DisplayCommand(this),
        )
    }

    override fun execute(context: CommandContext): Boolean {
        context.sender.sendMessage(
            Component.join(
                JoinConfiguration.noSeparators(),
                Component.text("Hopper Counter: ", NamedTextColor.GRAY),
                Component.text(
                    HopperCounter.isEnabled(),
                    if (HopperCounter.isEnabled()) NamedTextColor.AQUA else NamedTextColor.GRAY,
                ),
            ),
        )
        return true
    }

    fun hasPermission(sender: org.bukkit.command.CommandSender, vararg subcommand: String): Boolean =
        hasPermission(permBase, sender, *subcommand)
}